### PR TITLE
Deltaspike invocationhandler

### PR DESCRIPTION
**Pull request by jsight**
do not merge yet

### CODE REVIEW

File: deltaspike-invocationhandler-advanced/pom.xml [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+
+            <dependency>
+                <groupId>org.apache.deltaspike.core</groupId>
+                <artifactId>deltaspike-core-api</artifactId>
+                <version>${version.deltaspike}</version>
+            </dependency>
+            <dependency>
+                <groupId>org.apache.deltaspike.core</groupId>
+                <artifactId>deltaspike-core-impl</artifactId>
+                <version>${version.deltaspike}</version>
```
Comment (pmuir) [inline deltaspike-invocationhandler-advanced/pom.xml]: You should add a TODO that this needs removing once we update the BOM to use DeltaSpike 0.4

**Comment by pmuir** [comment]
This looks good, though I think we can probably just drop the more basic example.

**Comment by jsight** [comment]
I have added comments to indicate that these dependencies should be removed. Should I also remove the basic example? I know that the functionality is fairly basic (showing mostly the "how", and not the "why"), but I was thinking that it could be useful to demonstrate some of the corner cases in the simplest possible way (without involving JPA, etc).

**Comment by sgilda** [comment]
I test the advanced quickstart and got this error:

[ERROR] Failed to execute goal on project jboss-as-deltaspike-invocationhandler-advanced: Could not resolve dependencies for project org.jboss.as.quickstarts:jboss-as-deltaspike-invocationhandler-advanced:war:7.1.2-SNAPSHOT: The following artifacts could not be resolved: org.apache.deltaspike.core:deltaspike-core-api:jar:0.4-incubating-SNAPSHOT, org.apache.deltaspike.core:deltaspike-core-impl:jar:0.4-incubating-SNAPSHOT: Could not find artifact org.apache.deltaspike.core:deltaspike-core-api:jar:0.4-incubating-SNAPSHOT -> [Help 1]

I also had a problem running the QS Tools plugin:

[ERROR] Failed to execute goal on project jboss-as-deltaspike-invocationhandler-advanced: Could not resolve dependencies for project org.jboss.as.quickstarts:jboss-as-deltaspike-invocationhandler-advanced:war:7.1.2-SNAPSHOT: Could not find artifact org.apache.deltaspike.core:deltaspike-core-api:jar:0.4-incubating-SNAPSHOT -> [Help 1]

I'll investigate further.

**Comment by jsight** [comment]
Yes, if deltaspike 0.4 is still not out, that dependency would have to be
built manually.

**Comment by sgilda** [comment]
OK. Do you know when it will be out? We don't want to merge this until then. :-)

Will it run with what is out now? If so, maybe we should modify the POM files to use it and upgrade later?

**Comment by jsight** [comment]
Deltaspike 0.3 lacks the required classes.

The following thread implies that a 0.4 release is imminent:
http://markmail.org/message/wbadupz6kcxzlkez?q=list:org.apache.incubator.deltaspike-dev+order:date-backward#query:list%3Aorg.apache.incubator.deltaspike-dev%20order%3Adate-backward+page:1+mid:kox5iatoxgakngvh+state:results

**Comment by sgilda** [comment]
OK. Let me know when it's available and I'll test it. Thanks!

**Comment by sgilda** [comment]
@jsight : What is the status of this quickstart. Are you still waiting for DeltaSpike 0.4? I don't see anything to indicate it's out.

**Comment by jsight** [comment]
Yes, still waiting on DeltaSpike 0.4. IMO, it should have been released by
now, but it is an Apache project so we wait...

**Comment by sgilda** [comment]
Thanks for the update!

**Comment by jsight** [comment]
It looks like DeltaSpike 0.4 came out at the end of May.

**Comment by sgilda** [comment]
@jsight : You say DeltaSpike 0.4 cam out in May. Do you need to modify the POM files to reflect this change?

I tried changing `<version.deltaspike>0.4-incubating-SNAPSHOT</version.deltaspike>` to `<version.deltaspike>0.4.Final</version.deltaspike>`,  but got apparently was wrong. I got this error:

[ERROR] Failed to execute goal on project jboss-as-deltaspike-invocationhandler-basic: Could not resolve dependencies for project org.jboss.as.quickstarts:jboss-as-deltaspike-invocationhandler-basic:war:7.1.2-SNAPSHOT: The following artifacts could not be resolved: org.apache.deltaspike.core:deltaspike-core-api:jar:0.4.Final, org.apache.deltaspike.core:deltaspike-core-impl:jar:0.4.Final: Could not find artifact org.apache.deltaspike.core:deltaspike-core-api:jar:0.4.Final in central (http://repo.maven.apache.org/maven2) -> [Help 1]

Then I searched Maven Central and triied `<version.deltaspike>0.4</version.deltaspike>`. With that, I get this compile error:

[ERROR] COMPILATION ERROR : 
[INFO] -------------------------------------------------------------
[ERROR] /home/sgilda/GitRepos/quickstart-jdf/deltaspike-invocationhandler-basic/src/main/java/org/jboss/as/quickstart/deltaspike/invocationhandler/ExampleInvocationHandlerBinding.java:[25,66] error: package org.apache.deltaspike.core.api.invocationhandler.annotation does not exist

[ERROR] /home/sgilda/GitRepos/quickstart-jdf/deltaspike-invocationhandler-basic/src/main/java/org/jboss/as/quickstart/deltaspike/invocationhandler/ExampleInvocationHandlerBinding.java:[37,1] error: cannot find symbol

I think something needs updating here. :-)

File: deltaspike-invocationhandler-advanced/pom.xml [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+            to work with JBoss EAP 6. It requires you have access to the JBoss EAP 6 
+            maven repository. -->
+        <!-- <version.org.jboss.bom>1.0.4.Final-redhat-1</version.org.jboss.bom> -->
+
+        <!-- other plugin versions -->
+        <version.compiler.plugin>2.3.1</version.compiler.plugin>
+        <version.war.plugin>2.1.1</version.war.plugin>
+        <version.surefire.plugin>2.4.3</version.surefire.plugin>
+        <!-- TODO - This should be removed once 0.4 is available and included in JDF by default -->
+        <version.deltaspike>0.4-incubating-SNAPSHOT</version.deltaspike>
```
Comment (rafabene) [inline deltaspike-invocationhandler-advanced/pom.xml]: You should use 1.0.7.CR4 BOM which already have Deltaspike 0.4 and remove this property.

**Comment by jsight** [comment]
@rafabene - The 1.0.7.CR4 BOM does not appear to contain versions for the deltaspike-partial-bean-module-impl module. Could this be added?

**Comment by jsight** [comment]
Also, it appears that this quickstart will require some reworking, as the naming conventions used for these packages have changed.

**Comment by jsight** [comment]
I have pushed commits to modify this to match the latest DeltaSpike API. DeltaSpike now calls this "Partial Bean" instead of "InvocationHandler".

**Comment by sgilda** [comment]
I tested both of these quickstarts and they work. However, I am not able to run the QS Tool checker successfully against either of them. 

@rafabene : Do you know what would cause the following error: 

[INFO] Running Checker: BomVersionChecker
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 5.248s
[INFO] Finished at: Mon Jun 17 10:28:58 EDT 2013
[INFO] Final Memory: 32M/338M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.jboss.maven.plugins:maven-qstools-plugin:1.0.0.Final:check (default-cli) on project jboss-as-deltaspike-partialbean-advanced: An error has occurred in Quickstarts Checker Report report generation. org.codehaus.plexus.context.ContextException: Unable to resolve context key: stacks -> [Help 1]
[ERROR] 

When I run with the stacktrace switch, it shows this:

Caused by: org.apache.maven.reporting.MavenReportException: org.codehaus.plexus.context.ContextException: Unable to resolve context key: stacks
    at org.jboss.maven.plugins.qstools.QSCheckerReporter.executeReport(QSCheckerReporter.java:229)
    at org.apache.maven.reporting.AbstractMavenReport.generate(AbstractMavenReport.java:196)
    at org.apache.maven.reporting.AbstractMavenReport.execute(AbstractMavenReport.java:104)
    ... 21 more
Caused by: org.jboss.maven.plugins.qstools.QSCheckerException: org.codehaus.plexus.context.ContextException: Unable to resolve context key: stacks
    at org.jboss.maven.plugins.qstools.checkers.AbstractProjectChecker.check(AbstractProjectChecker.java:94)
    at org.jboss.maven.plugins.qstools.QSCheckerReporter.executeReport(QSCheckerReporter.java:216)
    ... 23 more
Caused by: org.codehaus.plexus.context.ContextException: Unable to resolve context key: stacks
    at org.codehaus.plexus.context.DefaultContext.get(DefaultContext.java:75)
    at org.jboss.maven.plugins.qstools.checkers.BomVersionChecker.processProject(BomVersionChecker.java:65)
    at org.jboss.maven.plugins.qstools.checkers.AbstractProjectChecker.check(AbstractProjectChecker.java:87)
    ... 24 more

**Comment by rafabene** [comment]
@jsight  ok. I'll add the module. REference: https://issues.jboss.org/browse/JDF-379

**Comment by sgilda** [comment]
@rafabene figured it out. I had a corrupt 0 byte length stacks cache file: /tmp/httpsrawgithubcomjbossjdfjdfstack100Finalstacksyamlstacks.yaml file. 

Once I removed it, the QS Tools works fine. 

I would like to see somthing logged to the server log as a result of each test, just to make it more interesting, but that's only a suggestion, not a requirement. :-)

@pmuir: I'm good with this. Do you need to review?

**Comment by rafabene** [comment]
-with-deltaspike BOM 1.0.7.CR5 released with deltaspike-partial-bean-module-impl and deltaspike-partial-bean-module-api

**Comment by jsight** [comment]
I have pushed an update using the 1.0.7.CR5 BOM.

On Mon, Jun 17, 2013 at 12:38 PM, Rafael Benevides <notifications@github.com

**Comment by sgilda** [comment]
These now build and run successfully!

@pmuir or @LightGuard : Do you need to do another review, or is this one good to go?

**Comment by pmuir** [comment]
Good to go :-)

**Comment by sgilda** [comment]
Squashed commits, rebased upstream, and merged.
